diff --git a/p2p/engine.cc b/p2p/engine.cc
index 39eeda3b..bb2a456e 100644
--- a/p2p/engine.cc
+++ b/p2p/engine.cc
@@ -1,4 +1,5 @@
 #include "engine.h"
+#include "adaptive_sleeper.h"
 #include "endpoint_wrapper.h"
 #include "util/debug.h"
 #include "util/pause.h"
@@ -2684,8 +2685,12 @@ void Endpoint::send_proxy_thread_func() {
   // bulk copy
   alignas(16) char task_buffer[16];
   UnifiedTask* task;
+  AdaptiveSleeper adaptive_sleeper;
+  adaptive_sleeper.update_timer();
 
   while (!stop_.load(std::memory_order_acquire)) {
+    adaptive_sleeper.maybe_sleep();
+
     if (jring_sc_dequeue_bulk(send_unified_task_ring_, task_buffer, 1,
                               nullptr) == 1) {
       task = *reinterpret_cast<UnifiedTask**>(task_buffer);
@@ -2726,6 +2731,7 @@ void Endpoint::send_proxy_thread_func() {
       auto* status = task->status_ptr;
       status->task_ptr.reset();
       status->done.store(true, std::memory_order_release);
+      adaptive_sleeper.update_timer();
     }
   }
 }
@@ -2736,8 +2742,11 @@ void Endpoint::recv_proxy_thread_func() {
   // bulk copy
   alignas(16) char task_buffer[16];
   UnifiedTask* task;
+  AdaptiveSleeper adaptive_sleeper;
+  adaptive_sleeper.update_timer();
 
   while (!stop_.load(std::memory_order_acquire)) {
+    adaptive_sleeper.maybe_sleep();
     if (jring_sc_dequeue_bulk(recv_unified_task_ring_, task_buffer, 1,
                               nullptr) == 1) {
       task = *reinterpret_cast<UnifiedTask**>(task_buffer);
@@ -2781,6 +2790,7 @@ void Endpoint::recv_proxy_thread_func() {
       auto* status = task->status_ptr;
       status->task_ptr.reset();
       status->done.store(true, std::memory_order_release);
+      adaptive_sleeper.update_timer();
     }
   }
 }
@@ -2819,8 +2829,15 @@ void Endpoint::ipc_poller_thread_func() {
   std::deque<IpcInflightOp*> active_ops;
   alignas(16) char buf[16];
   int cur_device = -1;
+  AdaptiveSleeper adaptive_sleeper;
+  adaptive_sleeper.update_timer();
 
   while (!stop_.load(std::memory_order_acquire) || !active_ops.empty()) {
+    // only try to sleep if there are no more inflight ops
+    if (active_ops.empty()) {
+      adaptive_sleeper.maybe_sleep();
+    }
+
     // Drain newly submitted ops into the local active list.
     while (jring_sc_dequeue_bulk(ipc_inflight_ring_, buf, 1, nullptr) == 1) {
       active_ops.push_back(*reinterpret_cast<IpcInflightOp**>(buf));
@@ -2865,6 +2882,7 @@ void Endpoint::ipc_poller_thread_func() {
       } else {
         ++it;
       }
+      adaptive_sleeper.update_timer();
     }
   }
 }
diff --git a/p2p/include/adaptive_sleeper.h b/p2p/include/adaptive_sleeper.h
new file mode 100644
index 00000000..35462455
--- /dev/null
+++ b/p2p/include/adaptive_sleeper.h
@@ -0,0 +1,89 @@
+#ifndef ADAPTIVE_SLEEPER_H
+#define ADAPTIVE_SLEEPER_H
+
+#include "util/debug.h"
+#include <chrono>
+#include <ctime>
+#include <sys/eventfd.h>
+#include <sys/poll.h>
+
+class AdaptiveSleeper {
+ public:
+  enum SleepState { POLL = 0, SLEEP };
+
+  AdaptiveSleeper()
+      : state_(POLL),
+        last_event_time_(std::chrono::steady_clock::time_point::min()) {
+    work_eventfd_ = eventfd(0, EFD_NONBLOCK);
+  }
+
+  ~AdaptiveSleeper() { close(work_eventfd_); };
+
+  void maybe_sleep() {
+    int ret;
+    if (std::chrono::steady_clock::now() - last_event_time_ >=
+            kNoActivityThreshold &&
+        last_event_time_ != std::chrono::steady_clock::time_point::min()) {
+      UCCL_LOG(INFO, UCCL_P2P) << "No activity detected for the last 120 "
+                                  "seconds, putting proxy to sleep";
+
+      state_ = SLEEP;
+    }
+
+    if (state_ == POLL) {
+      return;
+    } else {
+      UCCL_LOG(INFO, UCCL_P2P) << "Sleeping proxy thread";
+
+      // continuously sleep while there are no new work entries
+      struct pollfd events_to_poll[kNumActivitiesToPoll] = {
+          {.fd = work_eventfd_, .events = POLLIN},
+      };
+
+      int n = ppoll(events_to_poll, kNumActivitiesToPoll, &kPollSleepDuration,
+                    NULL);
+      UCCL_PCHECK(n != -1);
+
+      if (n > 0) {
+        if (events_to_poll[0].revents == POLLIN) {
+          UCCL_LOG(INFO, UCCL_P2P) << "Waking up becuase of work signal";
+          eventfd_t work_val;
+          ret = eventfd_read(work_eventfd_, &work_val);
+          UCCL_PCHECK(ret == 0);
+
+          last_event_time_ = std::chrono::steady_clock::now();
+          state_ = POLL;
+        }
+      }
+    }
+  }
+
+  void maybe_wake_proxy_thread() {
+    int ret = eventfd_write(work_eventfd_, kWakeEventConst);
+    UCCL_CHECK(ret == 0);
+  }
+
+  void update_timer() {
+    char const* is_adaptive_sleep = std::getenv("UCCL_RDMA_ADAPTIVE_SLEEP");
+    if (is_adaptive_sleep && strcmp(is_adaptive_sleep, "1") == 0) {
+      UCCL_LOG(INFO, UCCL_P2P) << "Adaptive sleeper configured";
+      last_event_time_ = std::chrono::steady_clock::now();
+    }
+  }
+
+ private:
+  // TODO: change after testing
+  static constexpr auto kNoActivityThreshold = std::chrono::seconds(3);
+  static constexpr int kNumActivitiesToPoll = 1;
+  static constexpr struct timespec kPollSleepDuration = {
+      .tv_sec = 5,
+      .tv_nsec = 0,
+  };
+  static constexpr int kWakeEventConst = 0x42;
+
+  SleepState state_;
+  std::chrono::steady_clock::time_point last_event_time_;
+  int work_eventfd_;
+};
+
+#endif
\ No newline at end of file
